single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,751 @@
+from single_kernel_mongo.events.tls import TLSEventsHandler
+from single_kernel_mongo.exceptions import (
+    ContainerNotReadyError,
+    SetPasswordError,
+    ShardingMigrationError,
+    UpgradeInProgressError,
+    WorkloadExecError,
+    WorkloadNotReadyError,
+    WorkloadServiceError,
+)
+from single_kernel_mongo.managers.backups import BackupManager
+from single_kernel_mongo.managers.config import (
+    CommonConfigManager,
+    LogRotateConfigManager,
+    MongoDBConfigManager,
+    MongoDBExporterConfigManager,
+    MongosConfigManager,
+)
+from single_kernel_mongo.managers.mongo import MongoManager
+from single_kernel_mongo.managers.tls import TLSManager
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.utils.mongo_connection import MongoConnection, NotReadyError
+from single_kernel_mongo.utils.mongodb_users import (
+    BackupUser,
+    MonitorUser,
+    OperatorUser,
+    get_user_from_username,
+)
+from single_kernel_mongo.workload import (
+    get_mongodb_workload_for_substrate,
+    get_mongos_workload_for_substrate,
+)
+from single_kernel_mongo.workload.mongodb_workload import MongoDBWorkload
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm  # pragma: nocover
+
+
+logger = logging.getLogger(__name__)
+
+
+@final
+class MongoDBOperator(OperatorProtocol, Object):
+    """Operator for MongoDB Related Charms."""
+
+    name = RoleEnum.MONGOD
+    workload: MongoDBWorkload
+
+    def __init__(self, charm: AbstractMongoCharm):
+        super(OperatorProtocol, self).__init__(charm, self.name)
+        self.charm = charm
+        self.substrate: Substrates = self.charm.substrate
+        self.role = ROLES[self.substrate][self.name]
+        self.state = CharmState(
+            self.charm,
+            self.substrate,
+            self.role,
+        )
+
+        container = (
+            self.charm.unit.get_container(CONTAINER) if self.substrate == Substrates.K8S else None
+        )
+
+        # Defined workloads and configs
+        self.define_workloads_and_config_managers(container)
+
+        # Managers
+        self.backup_manager = BackupManager(
+            self.charm,
+            self.role,
+            self.substrate,
+            self.state,
+            container,
+        )
+        self.tls_manager = TLSManager(
+            self,
+            self.workload,
+            self.state,
+            self.substrate,
+        )
+        self.mongo_manager = MongoManager(
+            self,
+            self.workload,
+            self.state,
+            self.substrate,
+        )
+
+        # Event Handlers
+        self.password_actions = PasswordActionEvents(self)
+        self.backup_events = BackupEventsHandler(self)
+        self.tls_events = TLSEventsHandler(self)
+        self.primary_events = PrimaryActionHandler(self)
+        self.client_events = DatabaseEventsHandler(self, RelationNames.DATABASE)
```

Comment:
It's good you raise this because I was wondering if I shouldn't put it into `single_kernel_mongo/operators`.
Please keep in mind also that those are not `charms` but operators.
If you want an example as I said, go to `tests/unit/mongodb_test_charm` to see how a charm is implemented in the end.

Reply:
maybe since they are operators instead of charms we could have an operators directory? along with a clear docstring explaining the difference between the two

Reply:
I'll bring that in the PR #5 with the other Operator and what it brings (there are a few changes as well)!